lib/charms/tempo_k8s/v1/charm_tracing.py

```diff
@@ -127,11 +131,29 @@ def get_current_span() -> Union[Span, None]:
     return cast(Span, span)
 
 
+def _get_tracer_from_context(ctx: Context) -> Optional[ContextVar]:
+    tracers = [v for v in ctx if v is not None and v.name == "tracer"]
+    if tracers:
+        return tracers[0]
+    return None
+
+
 def _get_tracer() -> Optional[Tracer]:
+    """Find tracer in context variable and as a fallback locate it in the full context."""
     try:
         return tracer.get()
     except LookupError:
-        return None
+        try:
```

Comment:
Why did you add this, and what is the use case? I don't remember having trouble using _get_tracer

Reply:
This is quite interesting. After the switch to the new dependency it seems there's been a change in the way `tracer` behaves in the context here. I started getting scenario test failures on `test_base_tracer_endpoint_event` so I started debugging and noticed that `_get_tracer()` returns None in the context of the span in the function instrumented by `@trace`. After going through the context I noticed that there is a `tracer` in the context but with a different reference, so I added this fallback retrieval.

To test it doesn't only happen on my local environment I've tried removing the fallback on a test draft PR and it looks the same behavior happens there: https://github.com/canonical/tempo-k8s-operator/actions/runs/7412918066/job/20170743398?pr=58

Branch diff: https://github.com/canonical/tempo-k8s-operator/compare/switch-to-otlp-http...temp/otlp-http-without-get-context

As fallback is triggered only in case `tracer.get()` doesn't find anything I felt it's okay to add it here in that way. I must admit though that it's dealing with a symptom, not addressing the cause I couldn't really pinpoint.

Do you think we should open a bug to investigate that further?

Reply:
yeah something doesn't feel right. Maybe the pydantic warnings are related? We bumped to >2, maybe that broke something somewhere? No need to fix it in this PR but let's open an issue to remind ourselves

Reply:
I see you created #60 for that, thanks!